codegen/smithy-aws-swift-codegen/src/test/kotlin/software/amazon/smithy/aws/swift/codegen/customizations/RulesBasedAuthSchemeResolverGeneratorTests.kt

```diff
@@ -0,0 +1,151 @@
+                            }
+                            sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: region)
+                            validAuthOptions.append(sigV4Option)
+                    }
+                    return validAuthOptions
+                }
+
+                public func constructParameters(context: HttpContext) throws -> ClientRuntime.AuthSchemeResolverParameters {
+                    return try DefaultS3AuthSchemeResolver().constructParameters(context: context)
+                }
+            }
+
+            public struct DefaultS3AuthSchemeResolver: S3AuthSchemeResolver {
+                public func resolveAuthScheme(params: ClientRuntime.AuthSchemeResolverParameters) throws -> [AuthOption] {
+                    var validAuthOptions = Array<AuthOption>()
+                    guard let serviceParams = params as? S3AuthSchemeResolverParameters else {
+                        throw ClientError.authError("Service specific auth scheme parameters type must be passed to auth scheme resolver.")
+                    }
+                    let endpointParams = EndpointParams(authSchemeParams: serviceParams)
+                    let endpoint = try DefaultEndpointResolver().resolve(params: endpointParams)
+                    guard let authSchemes = endpoint.authSchemes() else {
+                        return try InternalModeledS3AuthSchemeResolver().resolveAuthScheme(params: params)
+                    }
+                    let schemes = try authSchemes.map { (input) -> AWSClientRuntime.AuthScheme in try AWSClientRuntime.AuthScheme(from: input) }
+                    for scheme in schemes {
+                        switch scheme {
+                            case .sigV4(let param):
+                                var sigV4Option = AuthOption(schemeID: "aws.auth#sigv4")
+                                sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: param.signingName)
+                                sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: param.signingRegion)
+                                validAuthOptions.append(sigV4Option)
+                            case .sigV4A(let param):
+                                var sigV4Option = AuthOption(schemeID: "aws.auth#sigv4a")
+                                sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: param.signingName)
+                                sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: param.signingRegionSet?[0])
+                                validAuthOptions.append(sigV4Option)
+                            default:
+                                throw ClientError.authError("Unknown auth scheme name: \(scheme.name)")
+                        }
+                    }
+                    return validAuthOptions
+                }
```

Comment:
Nit: I would really prefer if this and similarly large/complex functions were broken up into helper functions but I know its difficult since we're code generating it. At the moment it is quite hard to read.

Reply:
In this case I think it's better to have the codegen code be easier to understand than the generated code be easier to understand, since what we're maintaining is the codegen code. To achieve good readability in generated output, the codegen code will be harder to understand as it creates multiple methods that work together in the end result.

Reply:
It doesnt have to be a tradeoff between good codegen code and good generated code we can achieve both! But I don't want to block your PR on this and since Josh doesn't have any issues here Ill leave it be